File: internal/mcpheaders/headers.go
// Package mcpheaders implements only the private SDK x-mcp-header behavior
// needed to validate a frozen public schema and regenerate downstream headers.
package mcpheaders

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"strconv"
	"strings"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

const (
	paramPrefix  = "Mcp-Param-"
	base64Prefix = "=?base64?"
	base64Suffix = "?="
	maxSafeInt   = 1<<53 - 1
)

type property struct {
	Type       string              `json:"type"`
	Annotation json.RawMessage     `json:"x-mcp-header,omitempty"`
	Properties map[string]property `json:"properties,omitempty"`
}

type binding struct {
	path   []string
	header string
}

// ValidateCall checks the applicable standard name and parameter headers.
func ValidateCall(header http.Header, name string, args json.RawMessage, tool *mcp.Tool) error {
	if got := header.Get("Mcp-Name"); got == "" || got != name {
		return fmt.Errorf("Mcp-Name does not match tool name")
	}
	bound, err := bindings(tool)
	if err != nil {
		return err
	}
	var values map[string]json.RawMessage
	if len(args) == 0 {
		args = []byte(`{}`)
	}
	if err := json.Unmarshal(args, &values); err != nil || values == nil {
		return fmt.Errorf("tool arguments must be an object")
	}
	for _, b := range bound {
		key := paramPrefix + b.header
		raw, present := lookup(values, b.path)
		headerValue := header.Get(key)
		if !present || string(raw) == "null" {
			if headerValue != "" {
				return fmt.Errorf("unexpected %s header", key)
			}
			continue
		}
		if headerValue == "" {
			return fmt.Errorf("%s does not match tool argument", key)
		}
		decoded, ok := decode(headerValue)
		if !ok {
			return fmt.Errorf("%s does not match tool argument", key)
		}
		value, ok := primitive(raw)
		if !ok || !value.equal(decoded) {
			return fmt.Errorf("%s does not match tool argument", key)
		}
	}
	return nil
}

// Generate produces exactly the applicable Mcp-Param-* values for args.
func Generate(tool *mcp.Tool, args json.RawMessage) (http.Header, error) {
	result := make(http.Header)
	if tool == nil {
		return nil, fmt.Errorf("missing tool schema")
	}
	var values map[string]json.RawMessage
	if len(args) == 0 {
		args = []byte(`{}`)
	}
	if err := json.Unmarshal(args, &values); err != nil || values == nil {
		return nil, fmt.Errorf("tool arguments must be an object")
	}
	bound, err := bindings(tool)
	if err != nil {
		return nil, err
	}
	for _, b := range bound {
		raw, ok := lookup(values, b.path)
		if !ok || string(raw) == "null" {
			continue
		}
		value, ok := primitive(raw)
		if !ok {
			return nil, fmt.Errorf("header-bound argument %q is not a supported primitive", strings.Join(b.path, "."))
		}
		result.Set(paramPrefix+b.header, encode(value.text))
	}
	return result, nil
}

type primitiveValue struct {
	text    string
	integer *big.Int
}

func (p primitiveValue) equal(header string) bool {
	if p.integer == nil {
		return header == p.text
	}
	other, ok := safeInteger(header)
	return ok && other.Cmp(p.integer) == 0
}

func primitive(raw json.RawMessage) (primitiveValue, bool) {
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.UseNumber()
	var value any
	if err := decoder.Decode(&value); err != nil {
		return primitiveValue{}, false
	}
	if err := decoder.Decode(new(any)); err != io.EOF {
		return primitiveValue{}, false
	}
	switch value := value.(type) {
	case string:
		return primitiveValue{text: value}, true
	case bool:
		return primitiveValue{text: strconv.FormatBool(value)}, true
	case json.Number:
		integer, ok := safeInteger(value.String())
		if !ok {
			return primitiveValue{}, false
		}
		return primitiveValue{text: integer.String(), integer: integer}, true
	default:
		return primitiveValue{}, false
	}
}

func safeInteger(value string) (*big.Int, bool) {
	rat, ok := new(big.Rat).SetString(value)
	if !ok || !rat.IsInt() {
		return nil, false
	}
	integer := rat.Num()
	maximum := big.NewInt(maxSafeInt)
	if integer.Cmp(maximum) > 0 || integer.Cmp(new(big.Int).Neg(maximum)) < 0 {
		return nil, false
	}
	return new(big.Int).Set(integer), true
}

func decode(value string) (string, bool) {
	if encoded, ok := strings.CutPrefix(value, base64Prefix); ok {
		if encoded, ok = strings.CutSuffix(encoded, base64Suffix); ok {
			decoded, err := base64.StdEncoding.DecodeString(encoded)
			if err != nil {
				return "", false
			}
			return string(decoded), true
		}
	}
	return value, true
}

func bindings(tool *mcp.Tool) ([]binding, error) {
	data, err := json.Marshal(tool.InputSchema)
	if err != nil {
		return nil, fmt.Errorf("encode input schema: %w", err)
	}
	var root property
	if json.Unmarshal(data, &root) != nil {
		return nil, fmt.Errorf("invalid input schema")
	}
	var out []binding
	seen := map[string]bool{}
	var walk func(map[string]property, []string) error
	walk = func(properties map[string]property, prefix []string) error {
		for name, p := range properties {
			path := append(append([]string(nil), prefix...), name)
			if p.Annotation != nil {
				var header string
				if json.Unmarshal(p.Annotation, &header) != nil || header == "" || !validHeader(header) || (p.Type != "string" && p.Type != "integer" && p.Type != "boolean") {
					return fmt.Errorf("invalid x-mcp-header annotation")
				}
				key := strings.ToLower(header)
				if seen[key] {
					return fmt.Errorf("duplicate x-mcp-header annotation")
				}
				seen[key] = true
				out = append(out, binding{path: path, header: header})
			}
			if err := walk(p.Properties, path); err != nil {
				return err
			}
		}
		return nil
	}
	if err := walk(root.Properties, nil); err != nil {
		return nil, err
	}
	return out, nil
}

func validHeader(value string) bool {
	for _, c := range value {
		if (c >= '0' && c <= '9') || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || strings.ContainsRune("!#$%&'*+-.^_`|~", c) {
			continue
		}
		return false
	}
	return value != ""
}

func lookup(values map[string]json.RawMessage, path []string) (json.RawMessage, bool) {
	var raw json.RawMessage
	var ok bool
	for i, part := range path {
		if i == 0 {
			raw, ok = values[part]
		} else {
			var nested map[string]json.RawMessage
			if json.Unmarshal(raw, &nested) != nil {
				return nil, false
			}
			raw, ok = nested[part]
		}
		if !ok {
			return nil, false
		}
	}
	return raw, true
}

func encode(value string) string {
	if value == "" {
		return value
	}
	unsafe := value[0] == ' ' || value[0] == '\t' || value[len(value)-1] == ' ' || value[len(value)-1] == '\t'
	for _, c := range value {
		unsafe = unsafe || c < 0x20 || c > 0x7e
	}
	if strings.HasPrefix(value, base64Prefix) && strings.HasSuffix(value, base64Suffix) {
		unsafe = true
	}
	if !unsafe {
		return value
	}
	return base64Prefix + base64.StdEncoding.EncodeToString([]byte(value)) + base64Suffix
}

File: src/conduit/headers.py
"""MCP tool x-mcp-header validation and invocation header generation."""

from __future__ import annotations

import base64
from collections.abc import Mapping
from decimal import Decimal, InvalidOperation
from typing import Any


class HeaderError(ValueError):
    """An annotated tool header cannot be safely generated."""


def validate_call_headers(tool: Mapping[str, Any], public_name: str, arguments: Mapping[str, Any], headers: Mapping[str, str]) -> None:
    if headers.get("mcp-name") != public_name:
        raise HeaderError("Mcp-Name must match tool name")
    generated = generate_headers(tool, arguments)
    lowered = {key.lower(): value for key, value in headers.items()}
    for key, expected in generated.items():
        actual = lowered.get(key.lower())
        if actual is None or not _equivalent(expected, _decode(actual)):
            raise HeaderError("Mcp-Param header does not match arguments")
    expected_names = {key.lower() for key in generated}
    if any(name.startswith("mcp-param-") and name not in expected_names for name in lowered):
        raise HeaderError("unexpected Mcp-Param header")


def generate_headers(tool: Mapping[str, Any], arguments: Mapping[str, Any]) -> dict[str, str]:
    schema = tool.get("inputSchema")
    if not isinstance(schema, Mapping):
        return {}
    output: dict[str, str] = {}
    seen: set[str] = set()
    _collect(schema.get("properties"), arguments, output, seen)
    return output


def _collect(properties: object, arguments: Mapping[str, Any], output: dict[str, str], seen: set[str]) -> None:
    if not isinstance(properties, Mapping):
        return
    for name, property_schema in properties.items():
        if not isinstance(name, str) or not isinstance(property_schema, Mapping):
            continue
        annotation = property_schema.get("x-mcp-header")
        if annotation is not None:
            if not isinstance(annotation, str) or not annotation or not _header_token(annotation):
                raise HeaderError("invalid x-mcp-header annotation")
            lowered = annotation.lower()
            if lowered in seen:
                raise HeaderError("duplicate x-mcp-header annotation")
            seen.add(lowered)
            if name in arguments:
                value = arguments[name]
                expected_type = property_schema.get("type")
                output["Mcp-Param-" + annotation] = _encode(expected_type, value)
        nested = arguments.get(name)
        _collect(property_schema.get("properties"), nested if isinstance(nested, Mapping) else {}, output, seen)


def _encode(expected_type: object, value: object) -> str:
    if expected_type == "string" and isinstance(value, str):
        return _wire_encode(value)
    if expected_type == "boolean" and isinstance(value, bool):
        return "true" if value else "false"
    if expected_type == "integer" and isinstance(value, (int, float)) and not isinstance(value, bool):
        number = Decimal(str(value))
        if number == number.to_integral_value() and -(2**53 - 1) <= number <= 2**53 - 1:
            return str(int(number))
    if expected_type not in {"string", "boolean", "integer"}:
        raise HeaderError("invalid x-mcp-header annotation type")
    raise HeaderError("invalid x-mcp-header argument")


def _header_token(value: str) -> bool:
    return all(char.isalnum() or char in "!#$%&'*+-.^_`|~" for char in value)


def _wire_encode(value: str) -> str:
    if not value:
        return value
    unsafe = value[0].isspace() or value[-1].isspace() or any(ord(char) < 0x20 or ord(char) > 0x7E for char in value)
    if value.startswith("=?base64?") and value.endswith("?="):
        unsafe = True
    return f"=?base64?{base64.b64encode(value.encode()).decode()}?=" if unsafe else value


def _decode(value: str) -> str:
    if value.startswith("=?base64?") and value.endswith("?="):
        try:
            return base64.b64decode(value[9:-2], validate=True).decode()
        except (ValueError, UnicodeDecodeError):
            return "\0invalid"
    return value


def _equivalent(expected: str, actual: str) -> bool:
    try:
        left, right = Decimal(expected), Decimal(actual)
    except InvalidOperation:
        return expected == actual
    if left != left.to_integral_value() or right != right.to_integral_value():
        return False
    maximum = Decimal(2**53 - 1)
    return -maximum <= left <= maximum and -maximum <= right <= maximum and left == right
